camus-api/src/main/java/com/linkedin/camus/schemaregistry/CachedSchemaRegistry.java

```
@@ -1,87 +1,155 @@
+    this.getByIdMinIntervalSeconds =
+        Integer.parseInt(props.getProperty(GET_SCHEMA_BY_ID_MIN_INTERVAL_SECONDS,
+            GET_SCHEMA_BY_ID_MIN_INTERVAL_SECONDS_DEFAULT));
+  }
+
+  public CachedSchemaRegistry(SchemaRegistry<S> registry, Properties props) {
+    this.registry = registry;
+    this.cachedById = new ConcurrentHashMap<CachedSchemaTuple, S>();
+    this.cachedLatest = new ConcurrentHashMap<String, SchemaDetails<S>>();
+    this.failedFetchHistories = new HashMap<String, FailedFetchHistory>();
+    this.init(props);
+  }
+
+  public String register(String topic, S schema) {
+    return registry.register(topic, schema);
+  }
+
+  public S getSchemaByID(String topic, String id) {
+    CachedSchemaTuple cacheKey = new CachedSchemaTuple(topic, id);
+    S schema = cachedById.get(cacheKey);
+    if (schema == null) {
+      synchronized (this) {
+        if (shouldFetchFromSchemaRegistry(id)) {
+          schema = fetchFromSchemaRegistry(topic, id);
+        } else {
+          throw new SchemaNotFoundException();
+        }
+      }
+    }
+    return schema;
+  }
+
+  private boolean shouldFetchFromSchemaRegistry(String id) {
+    if (!failedFetchHistories.containsKey(id)) {
+      return true;
+    }
+    FailedFetchHistory failedFetchHistory = failedFetchHistories.get(id);
+    boolean maxRetriesNotExceeded = failedFetchHistory.numOfAttempts < getByIdMaxRetries;
+    boolean minRetryIntervalSatisfied =
+        System.nanoTime() - failedFetchHistory.previousAttemptTime >= TimeUnit.SECONDS
+            .toNanos(getByIdMinIntervalSeconds);
+    return maxRetriesNotExceeded && minRetryIntervalSatisfied;
+  }
+
+  private S fetchFromSchemaRegistry(String topic, String id) {
+    try {
+      S schema = registry.getSchemaByID(topic, id);
+      return schema;
+    } catch (SchemaNotFoundException e) {
+      addFetchToFailureHistory(id);
+      throw new SchemaNotFoundException(e);
```

nit: Why not rethrow the same exception? this will add another meaningless "Caused by:" section.